### PR TITLE
Additional target ID stuff

### PR DESCRIPTION
**Pull request by TimGoll**
This should be the last PR concerning target ID TTT support. I added support for the rest of the TTT vanilla items.

Additionally I fixed some minor bugs, like:
- wrong initial zoomlevel for binoculars
- binoculars respect crosshairthickness
- DNA scanner respects crosshairthickness

### CODE REVIEW

File: gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua [hunk 1]
```diff
@@ -0,0 +1,829 @@
...
+	SWEP.PrintName = "dna_name"
+	SWEP.Slot = 8
+
+	SWEP.ViewModelFOV = 10
+	SWEP.DrawCrosshair = false
+
+	SWEP.EquipMenuData = {
+		type = "item_weapon",
+		desc = "dna_desc"
+	};
```
Comment (Alf21) [inline gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua]: `;` ?

File: gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua [hunk 2]
```diff
@@ -0,0 +1,829 @@
...
+					self:Report("dna_identify")
+				end
+			elseif ent.fingerprints and #ent.fingerprints > 0 then
+				self:GatherObjectSample(ent)
+			else
+				self:Report("dna_notfound")
+			end
+		end
+	else
+		if CLIENT then
```
Comment (Alf21) [inline gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua]: this if can be included to the previous scope (else to elseif)

File: gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua [hunk 3]
```diff
@@ -0,0 +1,829 @@
...
+			self:GetOwner():EmitSound(beep_miss)
```
Comment (Alf21) [inline gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua]: self:GetOwner can be localized

File: gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua [hunk 4]
```diff
@@ -0,0 +1,829 @@
...
+		end
+	end
+
+	self:GetOwner():LagCompensation(false)
+end
+
+function SWEP:GatherRagdollSample(ent)
+	local sample = ent.killer_sample or {t=0, killer=nil}
+	local ply = sample.killer
+	if (not IsValid(ply)) and sample.killer_sid then
```
Comment (Alf21) [inline gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua]: brackets are not necessary

File: gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua [hunk 5]
```diff
@@ -0,0 +1,829 @@
...
+		ply = player.GetBySteamID(sample.killer_sid)
+	end
+
+	if IsValid(ply) then
```
Comment (Alf21) [inline gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua]: scope could get minimized

File: gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua [hunk 6]
```diff
@@ -0,0 +1,829 @@
...
+	if #self.ItemSamples < self.MaxItemSamples then
+		local prnt = {
+			source = corpse,
+			ply = killer,
+			type = SAMPLE_PLAYER,
+			cls = killer:GetClass()
+		}
+
+		if not table.HasTable(self.ItemSamples, prnt) then
+			table.insert(self.ItemSamples, prnt)
```
Comment (Alf21) [inline gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua]: Do not use table,insert

File: gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua [hunk 7]
```diff
@@ -0,0 +1,829 @@
...
+
+			DamageLog("SAMPLE:\t " .. self:GetOwner():Nick() .. " retrieved DNA of " .. (IsValid(killer) and killer:Nick() or "<disconnected>") .. " from corpse of " .. (IsValid(corpse) and CORPSE.GetPlayerNick(corpse) or "<invalid>"))
```
Comment (Alf21) [inline gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua]: self:GetOwner() can be localized

File: gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua [hunk 8]
```diff
@@ -0,0 +1,829 @@
...
+		end
+
+		return true
+	end
+
+	return false
+end
+
+function SWEP:AddItemSample(ent)
+	if #self.ItemSamples < self.MaxItemSamples then
```
Comment (Alf21) [inline gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua]: scope could get minimized

File: gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua [hunk 9]
```diff
@@ -0,0 +1,829 @@
...
+		self:ClearScanState()
+	end
+
+	table.remove(self.ItemSamples, idx)
+	
+	self:SendPrints(false)
+end
+
+function SWEP:SecondaryAttack()
+	self:SetNextSecondaryFire( CurTime() + 0.05 )
```
Comment (Alf21) [inline gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua]: what is with these spaces?

File: gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua [hunk 10]
```diff
@@ -0,0 +1,829 @@
...
+if SERVER then
+	-- Sending this all in one umsg limits the max number of samples. 17 player
+	-- samples and 20 item samples (with 20 matches) has been verified as
+	-- working in the old DNA sampler.
+	function SWEP:SendPrints(should_open)
+		net.Start("TTT_ShowPrints", self:GetOwner())
+		net.WriteBit(should_open)
+		net.WriteUInt(#self.ItemSamples, 8)
+
+		for k, v in ipairs(self.ItemSamples) do
```
Comment (Alf21) [inline gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua]: ipairs can be replaced with for loop

File: gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua [hunk 11]
```diff
@@ -0,0 +1,829 @@
...
+			net.WriteString(v.cls)
+		end
+
+		net.Send(self:GetOwner())
+	end
+
+	function SWEP:SendScan(pos)
+		local clear = (pos == nil) or (not IsValid(self:GetOwner()))
```
Comment (Alf21) [inline gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua]: missing empty line

File: gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua [hunk 12]
```diff
@@ -0,0 +1,829 @@
...
+
+		if not IsValid(target) then return end
+
+		-- decoys always take priority, even after death
+		if IsValid(target.decoy) then
+			target = target.decoy
+		elseif not target:IsTerror() then
+			-- fall back to ragdoll, as long as it's not destroyed
+			target = target.server_ragdoll
+			if not IsValid(target) then return end
```
Comment (Alf21) [inline gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua]: missing empty line

File: gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua [hunk 13]
```diff
@@ -0,0 +1,829 @@
...
+		return target
+	end
+
+	function SWEP:PerformScan(idx, repeated)
+		if self:GetCharge() < MAX_CHARGE then return end
+
+		local sample = self.ItemSamples[idx]
+
+		if (not sample) or (not IsValid(self:GetOwner())) then
+			if repeated then self:ClearScanState() end
```
Comment (Alf21) [inline gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua]: bad format

File: gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua [hunk 14]
```diff
@@ -0,0 +1,829 @@
...
+
+		return target
+	end
+
+	function SWEP:PerformScan(idx, repeated)
+		if self:GetCharge() < MAX_CHARGE then return end
+
+		local sample = self.ItemSamples[idx]
+
+		if (not sample) or (not IsValid(self:GetOwner())) then
```
Comment (Alf21) [inline gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua]: brackets?

File: gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua [hunk 15]
```diff
@@ -0,0 +1,829 @@
...
+
+		local pos = target:LocalToWorld(target:OBBCenter())
+
+		self:SendScan(pos)
+		self:SetLastScanned(idx)
+		self.NowRepeating = self:GetRepeating()
+
+		local dist = math.ceil(self:GetOwner():GetPos():Distance(pos))
+
+		self:SetCharge(math.max(0, self:GetCharge() - math.max(50, dist / 2)))
```
Comment (Alf21) [inline gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua]: `x / 2` == `x * 0.5`

File: gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua [hunk 16]
```diff
@@ -0,0 +1,829 @@
...
+			end
+		end
+
+		return true
+	end
+end
+
+-- Helper to get at a player's scanner, if he has one
+local function GetTester(ply)
+	if IsValid(ply) then
```
Comment (Alf21) [inline gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua]: scope could get minimized

File: gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua [hunk 17]
```diff
@@ -0,0 +1,829 @@
...
+		self:DrawHelp()
+
+		local spos = self:GetOwner():GetShootPos()
+		local sdest = spos + (self:GetOwner():GetAimVector() * self.Range)
+
+		local tr = util.TraceLine({
+			start = spos,
+			endpos = sdest,
+			filter = self:GetOwner(),
+			mask=MASK_SHOT
```
Comment (Alf21) [inline gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua]: spaces?

File: gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua [hunk 18]
```diff
@@ -0,0 +1,829 @@
...
+		})
+		local ent = tr.Entity
+
+		local length = 20
+		local gap = 6
+		local thickness = math.floor(GetConVar("ttt_crosshair_thickness"):GetFloat())
```
Comment (Alf21) [inline gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua]: ConVars should be stored outside the function as var

File: gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua [hunk 19]
```diff
@@ -0,0 +1,829 @@
...
+		local offset = thickness * 0.5
+
+		local can_sample = false
+
+		if IsValid(ent) then
+			-- weapon or dropped equipment
+			if ((ent:IsWeapon() or ent.CanHavePrints) or
+			-- knife in corpse, or a ragdoll
+			ent:GetNWBool("HasPrints", false) or
```
Comment (Alf21) [inline gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua]: bad format

File: gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua [hunk 20]
```diff
@@ -0,0 +1,829 @@
...
+				can_sample = true
+			else
+				surface.SetDrawColor(255, 0, 0, 200)
+				gap = 0
+			end
+		else
+			surface.SetDrawColor(255, 255, 255, 200)
+		end
+
+		local x = ScrW() / 2.0
```
Comment (Alf21) [inline gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua]: / 2 ...

File: gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua [hunk 21]
```diff
@@ -0,0 +1,829 @@
...
+		local y = ScrH() / 2.0
```
Comment (Alf21) [inline gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua]: / 2 ...

File: gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua [hunk 22]
```diff
@@ -0,0 +1,829 @@
...
+				if IsValid(pnl) then
+					pnl:Remove()
+				end
+			end
+
+			s.Panels = {}
+
+			local last_scan = tester and tester:GetLastScanned() or -1
+
+			for k, v in ipairs(tbl) do
```
Comment (Alf21) [inline gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua]: ipairs

File: gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua [hunk 23]
```diff
@@ -0,0 +1,829 @@
...
+		delwrap:SetSize(370, bh)
+		delwrap:SetPaintBackground(false)
+
+		local delitem = vgui.Create("DButton", delwrap)
+		delitem:SetPos(0,0)
+		delitem:SetSize(bw, bh)
+		delitem:SetText(T("dna_menu_remove"))
+
+		delitem.DoClick = function()
+			if IsValid(ilist) and IsValid(ilist.SelectedPanel) then
```
Comment (Alf21) [inline gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua]: scope could get minimized

File: gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua [hunk 24]
```diff
@@ -0,0 +1,829 @@
...
+		state:SizeToContents()
+
+		local scan = vgui.Create("DButton", mwrap)
+		scan:SetText(T("dna_menu_scan"))
+		scan:SetSize(bw, bh)
+		scan:SetPos(0, 40)
+		scan:SetEnabled(false)
+
+		scan.DoClick = function(s)
+			if IsValid(ilist) then
```
Comment (Alf21) [inline gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua]: bad format

File: gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua [hunk 25]
```diff
@@ -0,0 +1,829 @@
...
+
+		-- Expose updating fns
+		dpanel.UpdatePrints = function(s, its)
+			if IsValid(iscroll) then
+				iscroll:LoadFrom(its)
+			end
+		end
+
+		dpanel.Think = function(s)
+			if IsValid(bar) and IsValid(scan) and tester then
```
Comment (Alf21) [inline gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua]: scope could get minimized

File: gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua [hunk 26]
```diff
@@ -0,0 +1,829 @@
...
+	local printspanel = nil
+
+	local function RecvPrints()
+		local should_open = net.ReadBit() == 1
+		local num = net.ReadUInt(8)
+		local item_prints = {}
+
+		for i=1, num do
+			local ent = net.ReadString()
+			table.insert(item_prints, ent)
```
Comment (Alf21) [inline gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua]: table.insert

File: gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua [hunk 27]
```diff
@@ -0,0 +1,829 @@
...
+	end
+
+	local printspanel = nil
+
+	local function RecvPrints()
+		local should_open = net.ReadBit() == 1
+		local num = net.ReadUInt(8)
+		local item_prints = {}
+
+		for i=1, num do
```
Comment (Alf21) [inline gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua]: missing spaces

File: gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua [hunk 28]
```diff
@@ -0,0 +1,829 @@
...
+
+	local beep_success = Sound("buttons/blip2.wav")
+
+	local function RecvScan()
+		local clear = net.ReadBit() == 1
+
+		if clear then
+			RADAR.samples = {}
+			RADAR.samples_count = 0
+			return
```
Comment (Alf21) [inline gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua]: ..

File: gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua [hunk 29]
```diff
@@ -0,0 +1,829 @@
...
+
+	local function RemoveSample(ply, cmd, args)
+		local tester = GetTester(ply)
+
+		if #args ~= 1 then return end
+
+		local idx = tonumber(args[1])
+
+		if not idx then return end
+		if not IsValid(tester) then return end
```
Comment (Alf21) [inline gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua]: combine ifs

File: gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua [hunk 30]
```diff
@@ -0,0 +1,829 @@
...
+	end
+end
+
+function SWEP:Reload()
+	return false
+end
+
+function SWEP:Deploy()
+	if SERVER and IsValid(self:GetOwner()) then
+		self:GetOwner():DrawViewModel(false)
```
Comment (Alf21) [inline gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua]: localize self:GetOwner()

File: gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua [hunk 31]
```diff
@@ -0,0 +1,829 @@
...
+	return true
+end
+
+if CLIENT then
+	function SWEP:DrawWorldModel()
+		if not IsValid(self:GetOwner()) then
+			self:DrawModel()
+		end
+	end
+end
```
Comment (Alf21) [inline gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua]: missing EOL

File: gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua (hunk 1 above)
Comment (TimGoll) [reply gamemodes/terrortown/entities/weapons/weapon_ttt_wtester.lua]: overseen this, was there from vanilla

**Pull request merged**